### PR TITLE
Font selector

### PR DESCRIPTION
**Pull request by spireken**
There's two demos for the Book View and Time and Sales window.

### CODE REVIEW

File: Applications/Spire/Source/Ui/FontSelectorButton.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  if(event->button() == Qt::LeftButton) {
+    swap_toggle();
+    m_clicked_signal();
+    update();
+  }
+}
+
+void FontSelectorButton::paintEvent(QPaintEvent* event) {
+  auto painter = QPainter(this);
+  if(m_is_toggled && !underMouse()) {
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/FontSelectorButton.cpp:59]:
```
if(underMouse()) {
  ...
} else if(m_is_toggled) {
  ...
} else {
  ...
}
```

File: Applications/Spire/Source/Ui/FontSelectorWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      font-size: %1px;
+    })").arg(scale_height(12)));
+  typeface_label->setFixedHeight(scale_height(14));
+  layout->addWidget(typeface_label);
+  auto grid_layout = new QGridLayout();
+  grid_layout->setHorizontalSpacing(HORIZONTAL_SPACING());
+  grid_layout->setVerticalSpacing(VERTICAL_SPACING());
+  layout->addLayout(grid_layout);
+  auto fonts = QFontDatabase().families();
+  auto font_list = [&] {
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/FontSelectorWidget.cpp:51]: Pre-allocate and `std::transform`.

File: Applications/Spire/Source/Ui/FontSelectorWidget.cpp [hunk 3]
```diff
@@ -48,13 +49,9 @@ FontSelectorWidget::FontSelectorWidget(const QFont& current_font,
...
-  auto font_list = [&] {
-    auto list = std::vector<QString>();
-    for(auto& font : fonts) {
-      list.push_back(font);
-    }
-    return list;
-  }();
+  auto font_list = std::vector<QString>();
+  font_list.reserve(fonts.size());
+  std::copy(fonts.begin(), fonts.end(), std::back_inserter(font_list));
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/FontSelectorWidget.cpp:54]: What about
`auto font_list = std::vector(fonts.begin(), fonts.end());`

File: Applications/Spire/Include/Spire/Ui/FontSelectorWidget.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#include "Spire/Ui/FontSelectorButton.hpp"
+#include "Spire/Ui/ValueInputWidget.hpp"
+
+namespace Spire {
+
+  //! Represents a widget for selecting system fonts.
+  class FontSelectorWidget : public QWidget {
+    public:
+
+      using FontPreviewSignal = Signal<void (const QFont& font)>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/FontSelectorWidget.hpp:15]: Document,

File: Applications/Spire/Include/Spire/Ui/FontSelectorWidget.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        \param font The initial font to display.
+        \param parent The parent widget.
+      */
+      explicit FontSelectorWidget(const QFont& font,
+        QWidget* parent = nullptr);
+
+      //! Returns the font represented by this widget.
+      const QFont& get_font() const;
+
+      boost::signals2::connection connect_font_preview_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/FontSelectorWidget.hpp:34]: Document.

File: Applications/Spire/Include/Spire/Ui/Ui.hpp [hunk 6]
```diff
@@ -5,6 +5,7 @@
 #include <QLineEdit>
 #include <QRect>
 #include <QSize>
+#include <QSpinBox>
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/Ui.hpp:8]: This header should likely be removed.

File: Applications/Spire/Include/Spire/Ui/ValueInputWidget.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#ifndef SPIRE_VALUE_INPUT_WIDGET_HPP
+#define SPIRE_VALUE_INPUT_WIDGET_HPP
+#include <QSpinBox>
+
+namespace Spire {
+
+  //! Represents a widget for inputting whole numbers.
+  class ValueInputWidget : public QSpinBox {
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/ValueInputWidget.hpp:8]: Rename this to something like ```IntegerInputWidget```, ```ValueInputWidget``` is too vague, everything is a value.

**Comment by spireken** [comment]
I'll fix these conflicts before merging.

**Pull request merged**